server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java

```diff
@@ -35,6 +35,60 @@
  * <p>
  * Similar to {@link ListenableActionFuture} and {@link ListenableFuture} except for its handling of exceptions: if this listener is
  * completed exceptionally then the exception is passed to subscribed listeners without modification.
+ * <p>
+ * Often this will be used to chain together a sequence of async actions such as in the following example:
+ * <pre>
+ * private void exampleAsyncMethod(String request, List&lt;Long> items, ActionListener&lt;Boolean> finalListener) {
+ *     SubscribableListener
+ *
+ *         // Start the chain and run the first step by creating a SubscribableListener using newForked():
+ *         .&lt;String>newForked(l -> firstAsyncStep(request, l))
+ *
+ *         // Run a second step when the first step completes using andThen(); if the first step fails then the exception falls through to
+ *         // the end without executing the intervening steps.
+ *         .&lt;Integer>andThen((l, firstStepResult) -> secondAsyncStep(request, firstStepResult, l))
+ *
+ *         // Run another step when the second step completes with another andThen() call; as above this only runs if the first two steps
+ *         // succeed.
+ *         .&lt;Boolean>andThen((l, secondStepResult) -> {
+ *             if (condition) {
+ *                 // Steps are exception-safe: an exception thrown here will be passed to the listener rather than escaping to the
+ *                 // caller.
+ *                 throw new IOException("failure");
+ *             }
+ *
+ *             // Steps can fan out to multiple subsidiary async actions using utilities like RefCountingListener.
+ *             final var result = new AtomicBoolean();
+ *             try (var listeners = new RefCountingListener(l.map(v -> result.get()))) {
+ *                 for (final var item : items) {
+ *                     thirdAsyncStep(secondStepResult, item, listeners.acquire());
+ *                 }
+ *             }
```

Comment:
Nit: L60-66 could be removed, I think, as it's standard to keep building listeners if desired and spelling it out here doesn't convey additional information about how to use SubscribableListener.

Reply:
I'd rather leave it here - it's a pretty common part of the pattern, and I've had conversations with other developers who didn't find this so obvious until they were shown it.